datafusion/core/src/datasource/mod.rs

```diff
@@ -199,3 +200,24 @@ fn get_col_stats(
         })
         .collect()
 }
+
+fn schema_eq_ignore_nullable(
```

Comment:
I could have sworn this function already existed (basically to compare two schemas ignoring nullabulity) but I could not find it

This is the closest I found: https://docs.rs/datafusion-common/26.0.0/datafusion_common/struct.DFSchema.html#method.equivalent_names_and_types but it is not quite the same

Reply:
we have the `nullify` internal function https://github.com/apache/arrow-datafusion/blob/d32f9a7425e6269bcc0b928a81f6d46871389bdf/datafusion/expr/src/logical_plan/builder.rs#L1042

and to compare 2 schemas ignoring nullability we using https://github.com/apache/arrow-rs/blob/15e0e76bfb6500799a43991f1339e69464c513f8/arrow-schema/src/fields.rs#L76

Reply:
UPD: https://github.com/apache/arrow-rs/blob/15e0e76bfb6500799a43991f1339e69464c513f8/arrow-schema/src/schema.rs#L323

Reply:
I think using `contains()` here may not be ideal :thinking: .
A non-nullable field does not `contains` a nullable field.

Reply:
@jonahgao That is true and depending on use and what schema is driving. I tested SqlLogic engine which currently uses `.contains` and found that
this query will fail

```
query I
select null a union all select 1 a;
----
```
and this does not
```query II
select 1 a union all select null a;
----
```
Depending on driving schema


Please try if `.contains` will work for you, if its not, we need to create a dedicated method similar to @alamb suggestion https://docs.rs/datafusion-common/26.0.0/datafusion_common/struct.DFSchema.html#method.equivalent_names_and_types but for `Schema`

If you create the method so I can also fix a hidden bug in sql logic test engine

Reply:
@comphead I have tried two possible ways of using it and both have failed.
1. Check `table_schema.contains(input_schema)`
The failed query is :
  ```sql
  DataFusion CLI v26.0.0
  ❯ create table t(a int not null);
  0 rows in set. Query took 0.027 seconds.
  
  ❯ insert into t values(1);
  Error during planning: Inserting query must have the same schema with the table.
  ```

2. Check `input_schema.contains(table_schema)`
The failed query is :
```sql
DataFusion CLI v26.0.0
❯ create table t2(a int null);
0 rows in set. Query took 0.029 seconds.
❯ create table t3(a int not null);
0 rows in set. Query took 0.002 seconds.

❯ insert into t2 select * from t3;
Error during planning: Inserting query must have the same schema with the table.
```

For an insertion operation, I think there is no need to have one schema be a superset of the other. 
Instead, we should make sure that there is an intersection between the data sets defined by the two schemas.

I've implemented the `equivalent_names_and_types` method for `Schema` in file `datafusion_common/dfschema.rs`. @alamb @comphead